src/ips/algorithm.rs

```diff
@@ -74,6 +74,23 @@ impl Ips {
 
     /// Generate peer list - main function with The Algorithm
     pub async fn generate(&mut self, state: &CrunchyState) -> Vec<Peer> {
+        // Sanity check that each node is really connected to its peers and the peers also
+        // have the node in their connections.
+        for (idx, node) in state.nodes.iter().enumerate() {
+            if node.connections.contains(&idx) {
+                println!("{} is connected to itself.", node.addr);
+            }
+
+            for peer in &node.connections {
+                if !state.nodes[*peer].connections.contains(&idx) {
+                    println!(
+                        "{} is not connected to {} but {} have a connection to it",
+                        node.addr, state.nodes[*peer].addr, node.addr
+                    );
+                }
+            }
+        }
```

Comment:
we only print these cases? nice!

Reply:
Yes, as talked on Slack thread, IPS should receive filtered input from such things. So, unless decision will change it's enough to just print information about such cases on input.

IPS prevents such situations on its output by:
- removing node itself from potential peer vector before making any selection;
- implementing this PR which fixed issue when only one side of the connection has been created 🥲  ;
- deleting duplicates from the final peerlist for each node.